src/ServiceControl.Monitoring.Transports.AzureServiceBus/QueueLengthProvider.cs

```diff
@@ -80,28 +97,52 @@ public Task Start()
             return TaskEx.Completed;
         }
 
-        async Task UpdateQueueLengthStore(ILookup<string, QueueDescription> queueData, CancellationToken token)
+        async Task<Dictionary<string, QueueDescription>> GetQueueList(CancellationToken token)
         {
-            var timestamp = DateTime.UtcNow.Ticks;
-            foreach (var mapping in endpointQueueMappings)
+            var pageSize = 100; //This is the maximal page size for GetQueueAsync
+            var pageNo = 0;
+
+            var queues = new List<QueueDescription>();
+
+            while (true)
             {
-                var queue = queueData[mapping.Value].FirstOrDefault();
-                if (queue != null)
+                var page = await managementClient.GetQueuesAsync(count:pageSize, skip: pageNo*pageSize, cancellationToken: token).ConfigureAwait(false);
+                
+                queues.AddRange(page);
+
+                if (page.Count < pageSize)
                 {
-                    var entries = new[]
-                    {
-                        new RawMessage.Entry
-                        {
-                            DateTicks = timestamp,
-                            Value = (await managementClient.GetQueueRuntimeInfoAsync(queue.Path, token).ConfigureAwait(false)).MessageCountDetails.ActiveMessageCount
-                        }
-                    };
-                    queueLengthStore.Store(entries, new EndpointInputQueue(mapping.Key.EndpointName, queue.Path));
+                    break;
                 }
-                else
+
+                pageNo++;
+            }
+
+            return queues.ToDictionary(q => q.Path, q => q);
+        }
+
+        Task UpdateAllQueueLengths(ConcurrentDictionary<string, QueueDescription> queues, CancellationToken token) => Task.WhenAll(endpointQueueMappings.Select(eq => UpdateQueueLength(eq, queues, token)));
```

Comment:
Do we know that this is safe? Is it okay to hit Azure with 1000 concurrent requests if you have 1000 queues?

Reply:
Good point, I have a pet project that hits NuGet concurrently and at about 20 concurrent calls I get throttled and things start to fail. Not the same service here but I think we should test this? (A little app to gen 1000 queues and then make the calls?)

Reply:
Did some testings. With the current implementation:
 1. We will grab all the queues from the namespace (paged, 100 queues at a time)
 2. We will ask of queue length only if we have nsb endpoint reporting to the sc. So the number of call per queue is equal to number of endpoints.

Test:
 1. 800 queues in the namespace and 200 nsb endpoints
 2. No errors/exceptions with the current approach. The execution of one query loop (queue list + length per queue) takes ~ 10 sec.

I propose to leave this as is. We can enhance this if customers need it.